File: Cpp/423.从英文中重建数字.cpp
/*
 * @lc app=leetcode.cn id=423 lang=cpp
 *
 * [423] 从英文中重建数字
 */

#include <iostream>
#include <string>
#include <unordered_map>
#include <vector>

using namespace std;

// @lc code=start
class Solution {
public:
    string originalDigits(string s) {
        unordered_map<char, int> c;
        for (char ch : s) {
            ++c[ch];
        }
        
        vector<int> cnt(10);
        cnt[0] = c['z'];
        cnt[2] = c['w'];
        cnt[4] = c['u'];
        cnt[5] = c['f'] - cnt[4];
        cnt[6] = c['x'];
        cnt[7] = c['s'] - cnt[6];
        cnt[8] = c['g'];
        cnt[1] = c['o'] - cnt[0] - cnt[2] - cnt[4];
        cnt[3] = c['h'] - cnt[8];
        cnt[9] = c['i'] - cnt[5] - cnt[6] - cnt[8];
        
        string ans;
        for (int i = 0; i < 10; ++i) {
            for (int j = 0; j < cnt[i]; ++j) {
                ans += char(i + '0');
            }
        }
        return ans;
    }
};
// @lc code=end

int main(int argc, const char * argv[]) {
    // insert code here...
    
    string s;
    string res;
    
    /**
     输入：s = "owoztneoer"
     输出："012"
     */
    s = "owoztneoer";
    res = "012";
    assert(res == Solution().originalDigits(s));
    
    /**
     输入：s = "fviefuro"
     输出："45"
     */
    s = "fviefuro";
    res = "45";
    assert(res == Solution().originalDigits(s));
    
    /**
     输入：s = "nnei"
     输出："9"
     */
    s = "nnei";
    res = "9";
    assert(res == Solution().originalDigits(s));

    cout << "OK~" << endl;

    return 0;
}

File: Swift/423.从英文中重建数字.swift
/*
 * @lc app=leetcode.cn id=423 lang=swift
 *
 * [423] 从英文中重建数字
 */

// @lc code=start
class Solution {
    func originalDigits(_ s: String) -> String {
        var c: [Character: Int] = [:]
        for ch in s {
            c[ch] = (c[ch] ?? 0) + 1
        }
        
        var cnt: [Int] = .init(repeating: 0, count: 10)
        
        cnt[0] = c["z"] ?? 0
        cnt[2] = c["w"] ?? 0
        cnt[4] = c["u"] ?? 0
        cnt[5] = (c["f"] ?? 0) - cnt[4]
        cnt[6] = (c["x"] ?? 0)
        cnt[7] = (c["s"] ?? 0) - cnt[6]
        cnt[8] = (c["g"] ?? 0)
        cnt[1] = (c["o"] ?? 0) - cnt[0] - cnt[2] - cnt[4]
        cnt[3] = (c["h"] ?? 0) - cnt[8]
        cnt[9] = (c["i"] ?? 0) - cnt[5] - cnt[6] - cnt[8]

        var ans: String = ""
        for i in 0..<10 {
            for _ in 0..<cnt[i] {
                ans += String(Character(UnicodeScalar(i + 48)!))
            }
        }
        return ans
    }
}
// @lc code=end

func main() {
    var s: String
    var res: String
    
    /**
     输入：s = "owoztneoer"
     输出："012"
     */
    s = "owoztneoer"
    res = "012"
    assert(res == Solution().originalDigits(s))
    
    /**
     输入：s = "fviefuro"
     输出："45"
     */
    s = "fviefuro"
    res = "45"
    assert(res == Solution().originalDigits(s))
    
    /**
     输入：s = "nnei"
     输出："9"
     */
    s = "nnei"
    res = "9"
    assert(res == Solution().originalDigits(s))
}
